File: 86-partition-list/partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode small = new ListNode(-1);
        ListNode head_small = small;
        ListNode big = new ListNode(-1);
        ListNode head_big = big;
        while(head!=null){
            if(head.val < x ){
                small.next = new ListNode(head.val);
                small = small.next;
                head = head.next;
            }
            else{
                big.next = new ListNode(head.val);
                big = big.next;
                head = head.next;
            }
        }
        small.next = head_big.next;
        return head_small.next;   
    }
}

File: 86-partition-list/partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* x1= new ListNode(-1);
        ListNode* x2= new ListNode(-1);
        ListNode* dummy1=x1;
        ListNode* dummy2=x2;
        while(head!=NULL){
            if(head->val<x){
                dummy1->next=head;
                dummy1=dummy1->next;
            }
            else{
                dummy2->next=head;
                dummy2=dummy2->next;
            }
            head=head->next;
        }
        dummy1->next=x2->next;
        dummy2->next=NULL;
        return x1->next;
    }
};